Sanitize JoinPath handling

feichai0017 opened the pull request:
Improve `FileSystem::JoinPath` to normalize separators and dot segments, respect absolute RHS, preserve URI schemes, and handle drive prefixes. 
Added file system tests covering separator normalization, .. folding, absolute override, and repeated separator cleanup.

feichai0017 committed:
Sanitize JoinPath handling and add tests

feichai0017 committed:
fix code format

Mytherin commented on src/common/file_system.cpp:
```diff
@@ -295,8 +295,120 @@ string FileSystem::GetWorkingDirectory() {
+			return false;
+		}
+		// avoid treating a Windows drive ("C:\") as a scheme
+		if (scheme_pos == 1 && StringUtil::CharacterIsAlpha(path[0])) {
```
Is this necessary? We are looking for `scheme://` above, are we treating `C://` as a Windows drive? Shouldn't that be only `C:\`?

Mytherin commented:
Thanks for the clarification - adding the SQL function in a follow-up makes sense to me.
Could you just add some more tests then, there are a number of edge cases mentioned in the code that appear to be untested, e.g. some tests I could think of:

* `JoinPath("dir/subdir/..", "sibling")`
* `JoinPath("./dir/./subdir/./..", "./sibling/.")`
* `JoinPath("dir/..", "../..")`
* `JoinPath("/usr/local", "../..")` (this should return "/" I think?)
* `JoinPath("/", "usr/local")` (follow-up from the previous one)
* `JoinPath("/usr/local", "../../..")` (not sure what this should return)
* `JoinPath("file:/usr/local", "../bin")`
* `JoinPath("file:/usr/local", "../bin")`
* `JoinPath("C:\", "..")`
* `JoinPath("C:\", "..\..")`
* `JoinPath("C:", "system32")`
* `JoinPath("C:\", "system32")`
* `JoinPath("C:drive_relative_path", "path")`
* `JoinPath("C:drive_relative_path", "..\path")` (I think this needs to be `C:path`, **not** `C:\path`, otherwise the relative property is lost)

Maybe we could also add some special tests for Windows specifically, i.e. when the separator is `\`.

feichai0017 committed:
Harden JoinPath normalization and edge tests

feichai0017 committed:
Fix Windows drive root joins and add glob regression test

feichai0017 committed:
correct coede format

feichai0017 committed:
Add path_join SQL function

feichai0017 committed:
fix code format

feichai0017 committed:
Regenerate function metadata for path_join

feichai0017 committed:
Make path_join SQL test platform-neutral

feichai0017 committed:
Stabilize path_join evaluation

benfleis commented on src/include/duckdb/function/scalar/string_functions.hpp:
```diff
@@ -375,6 +375,16 @@ struct RegexpEscapeFun {
+struct PathJoinFun {
+	static constexpr const char *Name = "path_join";
+	static constexpr const char *Parameters = "path,...";
+	static constexpr const char *Description = "Joins path components using the platform separator, normalizing duplicate separators and dot segments. An absolute RHS overrides previous components.";
```
I think this needs update -- absolute RHS now only overrides if a child, and no mention of strict mode in description, params, examples, etc.

benfleis reviewed (changes requested):
@feichai0017 The bulk of the PR looks great, just a couple small tweaks and should be ready to go, thank you again! I'm literally waiting to incorporate this into my another fix so I'll help get it through right away :)

feichai0017 commented on src/function/scalar/string/path_join.cpp:
```diff
@@ -0,0 +1,146 @@
+	explicit PathJoinBindData(bool strict_p, idx_t path_count_p) : strict(strict_p), path_count(path_count_p) {
+	}
+
+	bool strict;
```
👌

feichai0017 commented:
@benfleis BTW, when I changed the path join fucntion to always throw an InValidInput exception, I tried TRY and found that it couldn't catch runtime exceptions unless I modified the expression to prevent TRY from being folded

feichai0017 committed:
Fix JoinPath file scheme handling

feichai0017 committed:
Relax drive-relative clamp and tidy separator stripping

feichai0017 committed:
Keep URI paths using forward slashes when joining

feichai0017 committed:
Revert "Keep URI paths using forward slashes when joining"

This reverts commit 9b06a6250e3548d09abf4bb8bac8dcbe825d0434.

feichai0017 committed:
Normalize URI paths to forward slashes in JoinPath

feichai0017 committed:
Normalize URI paths to forward slashes when separator is slash

feichai0017 committed:
Force forward slashes for URI joins regardless of platform separator

feichai0017 committed:
Fix JoinPath URI separator handling

feichai0017 committed:
Fix file scheme joins with absolute paths

feichai0017 committed:
Clarify path_join strictness and Windows path handling

feichai0017 committed:
Format path_join description

feichai0017 committed:
Fix JoinPath root separator duplication

feichai0017 committed:
correct code format

feichai0017 committed:
update function generation

benfleis reviewed (approved):
LGTM! @lnkuiper incoming

lnkuiper commented on src/function/scalar/string/path_join.cpp:
```diff
@@ -0,0 +1,110 @@
+	idx_t path_count = arguments.size();
+	vector<LogicalType> arg_types(arguments.size(), LogicalType::VARCHAR);
+	bound_function.arguments = std::move(arg_types);
+	bound_function.null_handling = FunctionNullHandling::SPECIAL_HANDLING;
```
I think the null handling of this function should be `DEFAULT_NULL_HANDLING`. If any of the inputs is `NULL`, the output should be `NULL` instead of throwing an exception.

lnkuiper reviewed (changes requested):
Thanks for the changes! I have two more comments, then I think this is ready to go:

feichai0017 committed:
Merge branch 'main' into sanitize-joinpath

feichai0017 committed:
Adjust path_join null handling and tests

feichai0017 committed:
Fix Windows glob path joining for drive roots

benfleis commented:
@feichai0017 ok, I pulled your latest changes (thank you!) into the PR I had worked on, which itself includes both a few new tests and 1 fix for an edge case (`/foo/bar://baz`), this can be found at #20292 and includes all of your changes.

I had done a squash of all your changes to keep the history cleaner, but the pull/merge seems to have re-expanded them. In any case, I'll send this on to be landed, as everything should be addressed now.
